src/type-layout.md

```diff
@@ -56,12 +56,23 @@ Most primitives are generally aligned to their size, although this is
 platform-specific behavior. In particular, on x86 u64 and f64 are only
 aligned to 32 bits.
 
+For the primitive numeric types (`u8`, `i8`, `u16`, `i16`, `u32`, `i32`, `u64`,
+`i64`, `u128`, `i128`, `usize`, `isize`, `f32`, and `f64`), every bit pattern
+represents a valid instance of the type (in other words, for every primitive numeric
+type, `T`, `transmute::<[u8; size_of::<T>()], T>(...)` is always sound). For the
+primitive numeric types and also for `bool` and `char`, every byte is guaranteed to be
+initialized (in other words, for every such type, `T`,
+`transmute::<T, [u8; size_of::<T>()]>(...)` is always sound).
+
 ## Pointers and References Layout
 
 Pointers and references have the same layout. Mutability of the pointer or
 reference does not change the layout.
 
-Pointers to sized types have the same size and alignment as `usize`.
+Pointers to sized types have the same size and alignment as `usize`. Every
+byte of a pointer to a sized type and of a reference to a sized type is
+initialized (in other words, for such a pointer or reference type, `P`,
+`transmute::<P, [u8; size_of::<P>()]>(...)` is always sound).
```

Comment:
Ah this is where ptr2int transmute is even directly mentioned. That's a huge topic for discussion, see https://github.com/rust-lang/unsafe-code-guidelines/issues/286. And it's far from an easy topic, see [this blog post](https://www.ralfj.de/blog/2022/04/11/provenance-exposed.html) for why it is tricky.

It is almost certainly the case that transmuting a pointer to usize and back will *not* give you back the original pointer, but something more like `ptr::invalid(ptr.addr())`. So even if we are willing to commit to ptr2int transmute not being insta-UB, we should caution against the caveats that come with it.

Reply:
Do you have a sense of what we _can_ guarantee about ptr2int or int2ptr? If the answer is "nothing", I'd be happy to just remove this from the PR.

Reply:
For ptr2int transmutes I don't think we are ready to guarantee anything yet (except for the case that is covered by https://github.com/rust-lang/rust/pull/113510).

int2ptr transmutes... I think we can commit to them being legal, but I don't know if we even can commit to the resulting pointer being valid for zero-sized accesses (Cc https://github.com/rust-lang/opsem-team/issues/12).

Reply:
Yeah I'd be fine just removing this paragraph entirely.